docs/source/notes/autograd.rst

```diff
@@ -3,119 +3,394 @@
+
+.. note::
+
+    - Both the `detach()` method and the `torch.no_grad` decorator are expected not to follow these rules.
+    - This is work in progress, if you find a :code:`Function` that does not follow these rules, please open a new issue with the tag "module:autograd".
+
+#. If the function is not defined (:math:`sqrt(-1)` or :math:`log(-1)` for example) then this is undefined. Most function will return :math:`nan`, but for performance reasons, some functions will return not-nan values (:math:`log(-1)` for example).
+#. If a gradient exist at the current point, return it.
+#. If the function is convex, return one of it's subgradients at the current point, if more than one exists, return the subgradient with minimum norm (As it is always a descent direction, see "Convex Optimization Algorithms" by Dimitri Bertsekas, Proposition 4.3.1 for a proof).
+#. Define the gradient at the current point by continuity (note that :math:`inf` is possible here). If multiple values are possible, pick one arbitrarily.
+
+.. note::
+
+    Even though we try our best, unless all the used elementary operations used are in case 2, the returned gradient is not guaranteed to be correct.
+
+
+Implementation details
+^^^^^^^^^^^^^^^^^^^^^^
+
+.. warn::
+
+  All the following elements are implementation details.
+  You should not rely on them being fixed but they can be useful to debug code that uses the current version.
+
+Computational graph
+"""""""""""""""""""
+
+To know in which order the chain rule should be applied, we need to record the order the operations are applied to be able to replay it in the reverse order later.
+This is known as tape-based automatic differentiation.
+In our case, we create a directed acyclic graph.
+The nodes, called :code:`Node` represent the backward computations to be performed for a give elementary operation.
+The edges link a given :code:`Node` to all the :code:`Node`s that created the inputs of the elementary operation it represents.
+
+You can access the :code:`Node` that created a given Tensor (if the Tensor requires gradients and is not a leaf Tensor) using the ``.grad_fn`` field.
+You can then access the edges that link the different :code:`Node`s by using the ``.next_functions`` field of a :code:`Node`.
+A handy package to explore this graph is [torchviz](https://github.com/szagoruyko/pytorchviz).
+
+We use ``autograd.Function`` as a nice interface to create this graph.
+Indeed, during the ``.apply``, the forward is invoked and all the inputs and outputs are connected properly to the exiting graph associated to the inputs.
+The :code:`Node` that is created is a wrapper around the user-defined backward function (and the name is appended with ``Backward`` as you can see in the
+graph examples in torchviz.
+
+
+Not Gradients
+"""""""""""""
+
+The operations that do not compute gradients that are presented above are implemented as follows to obtain the behavior described above.
```

Comment:
It'd be nice to have a code snippet to illustrate what comes below.

Reply:
I'm not sure what to put as code sample here :/ Do you have something in mind?

Reply:
Can you refer to examples done above then?